### PR TITLE
external repo: checkout revision before initializing dvc repo

### LINKED ISSUE #2848: dvc get from dvc-branch in not dvc-repo
```
$ dvc -V
0.70.0+38be14
```

`master` branch in https://github.com/dmpetrov/test_repo is NOT a DVC-repo. However, the extracted branch `b-test` IS DVC-repo and should work:

```
$ dvc get https://github.com/dmpetrov/test_repo --rev  b-test b-file
ERROR: failed to get 'b-file' from 'https://github.com/dmpetrov/test_repo' - URL 'https://github.com/dmpetrov/test_repo' is not a dvc repository.
```

This a regression since this command worked before: `dvc get https://github.com/topusOctopus/mmdetection/ --rev fasterRCNN_resnet50_support model_weights/faster_rcnn_r50_c4_2x-6e4fdf4f.pth`

### PR DESCRIPTION
**Pull request by pared**
* [x] ❗ Have you followed the guidelines in the [Contributing to DVC](https://dvc.org/doc/user-guide/contributing/core) list?

* [x] 📖 Check this box if this PR **does not** require [documentation](https://dvc.org/doc) updates, or if it does **and** you have created a separate PR in [dvc.org](https://github.com/iterative/dvc.org) with such updates (or at least opened an issue about it in that repo). Please link below to your PR (or issue) in the [dvc.org](https://github.com/iterative/dvc.org) repo.

* [x] ❌ Have you checked DeepSource, CodeClimate, and other sanity checks below? We consider their findings recommendatory and don't expect everything to be addressed. Please review them carefully and fix those that actually improve code or fix bugs.

Thank you for the contribution - we'll try to review it as soon as possible. 🙏

Fixes #2848

### CODE REVIEW

**Comment by pared** [comment]
If this change gets accepted we should consider replacing `repo_dir` with `empty_dir` some tests. It seems to me that `repo_dir` has been used as empty dir initializer at least in [some cases](https://github.com/iterative/dvc/blob/38be142497a03feb5e23155e8aed089c59dec19e/tests/func/test_get.py#L28). Using `empty_dir` will make potential test debuggin easier.

File: tests/func/test_get.py [hunk 1]
```diff
@@ -87,3 +88,33 @@ def test_get_to_dir(dname, repo_dir, erepo):
...
+    git_erepo.git.git.checkout("master", b=dvc_branch)
+    git_erepo.dvc_branch = dvc_branch
+
+    dvc_repo = Repo.init(git_erepo._root_dir)
+    stage, = dvc_repo.add([git_erepo.FOO])
+    dvc_repo.scm.add([".dvc", stage.relpath])
+
+    rconfig = RemoteConfig(dvc_repo.config)
+    rconfig.add("upstream", dvc_repo.cache.local.cache_dir, default=True)
+    dvc_repo.scm.add([dvc_repo.config.config_file])
```
Comment (pared) [inline tests/func/test_get.py]: Can be removed after #2780

File: tests/func/test_get.py [hunk 2]
```diff
@@ -87,3 +88,33 @@ def test_get_to_dir(dname, repo_dir, erepo):
...
+    yield git_erepo
+
+
+def test_get_from_non_dvc_master(empty_dir, erepo_no_dvc_master):
+    Repo.get(
+        erepo_no_dvc_master._root_dir,
+        "foo",
+        out="foo",
+        rev=erepo_no_dvc_master.dvc_branch,
+    )
```
Comment (Suor) [inline tests/func/test_get.py]: We don't check that it works.

File: tests/func/test_get.py (hunk 1 above)
Comment (Suor) [reply tests/func/test_get.py]: We only use it once. Why not just inline it into the test?

File: dvc/external_repo.py [hunk 3]
```diff
@@ -61,11 +61,12 @@ def _external_repo(url=None, rev=None, cache_dir=None):
         REPO_CACHE[url, None, None] = clean_clone_path
 
     # Adjust new clone/copy to fit rev and cache_dir
+
+    # Checkout first in case of non dvc master
+    checkout_revision(new_path, rev)
+
     repo = Repo(new_path)
```
[the remark this answers is not in the document]
Comment (efiop) [reply dvc/external_repo.py]: @Suor I guess "# Adjust new clone/copy to fit rev and cache_dir" is for this, though it looks weird detached like that now.

[the remark this answers is not in the document]
Comment (pared) [reply dvc/external_repo.py]: done

File: tests/basic_env.py [hunk 4]
```diff
@@ -21,7 +21,55 @@
 logger = logging.getLogger("dvc")
 
 
-class TestDirFixture(object):
+class EmptyDirFixture(object):
```
Comment (mroutis) [inline tests/basic_env.py]: @pared, not sure if it is the same, but what about?

```python
@pytest.fixture
def chdir_tmp(tmp_path):
    old_directory = os.cwd()
    new_directory = fspath_py35(tmp_path)

    os.chdir(new_directory)
    yield
    os.chdir(old_directory)
```

Also, why refactoring something that we will deprecate eventually) ?

File: dvc/external_repo.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (efiop) [reply dvc/external_repo.py]: @pared i suppose you didn't push yet?

File: tests/conftest.py [hunk 5]
```diff
@@ -190,3 +190,11 @@ def git_erepo():
     repo.setUp()
     yield repo
     repo.tearDown()
+
+
+@pytest.fixture
+def empty_dir():
```
Comment (mroutis) [inline tests/conftest.py]: see this comment: https://github.com/iterative/dvc/pull/2852/files#r352164905

File: dvc/external_repo.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (Suor) [reply dvc/external_repo.py]: But why do you need to checkout first?

[the remark this answers is not in the document]
Comment (Suor) [reply dvc/external_repo.py]: Also, logically it should be:
```python
if rev is not None:
    _git_checkout(new_path, rev)
```
To stay in line with the rest.

File: tests/conftest.py (hunk 5 above)
Comment (Suor) [reply tests/conftest.py]: This won't be needed when new dir_helper fixtures land. So maybe save effort on this refactoring. If you simply need an empty dir you may use `tmp_path` pytest standard fixture and `monkeypatch` to chdir to it:
```python
@pytest.fixture
def empty_dir(tmp_path, monkeypatch):
    monkeypatch.chdir(tmp_path)
    return tmp_path

```

File: tests/func/test_get.py [hunk 6]
```diff
@@ -87,3 +89,38 @@ def test_get_to_dir(dname, erepo):
 
     assert os.path.isdir(dname)
     assert filecmp.cmp(erepo.FOO, dst, shallow=False)
+
+
+def test_get_from_non_dvc_master(empty_dir, git_erepo, caplog):
```
Comment (Suor) [inline tests/func/test_get.py]: How about using `erepo` just removing `.dvc` dir at the start of the test?

File: dvc/external_repo.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (pared) [reply dvc/external_repo.py]: done

File: dvc/external_repo.py [hunk 7]
```diff
@@ -61,11 +61,12 @@ def _external_repo(url=None, rev=None, cache_dir=None):
...
     try:
-        if rev is not None:
-            repo.scm.checkout(rev)
-
         if cache_dir is not None:
             cache_config = CacheConfig(repo.config)
             cache_config.set_dir(cache_dir, level=Config.LEVEL_LOCAL)
```
[the remark this answers is not in the document]
Comment (pared) [reply dvc/external_repo.py]: I agree that it can be done with just config, though it would require a change in `test_get_from_non_dvc_repo`. I'll create an issue for that.

[the remark this answers is not in the document]
Comment (pared) [reply dvc/external_repo.py]: https://github.com/iterative/dvc/issues/2880

File: tests/func/test_get.py (hunk 6 above)
Comment (pared) [reply tests/func/test_get.py]: I can do that, but test won't be much more readable due to fact that the default remote set inside `erepo` creation is its own cache and I need to provide a new one during this particular test.

Comment (pared) [reply tests/func/test_get.py]: After giving it some thought, that is not possible. Such a case would defeat the test purpose. External repo is cloning the target, so it doesn't matter that I remove `.dvc` from `erepo`, it will still be `dvc` repo after clone.

File: dvc/external_repo.py [hunk 8]
```diff
@@ -61,11 +61,14 @@ def _external_repo(url=None, rev=None, cache_dir=None):
         REPO_CACHE[url, None, None] = clean_clone_path
 
     # Adjust new clone/copy to fit rev and cache_dir
+
+    # Checkout needs to be done first because current branch might not be dvc
+    # DVC repository
```
Comment (efiop) [inline dvc/external_repo.py]: "dvc DVC" looks like typo.

Comment (pared) [reply dvc/external_repo.py]: done

**Comment by efiop** [comment]
@pared check the tests, lots of them are failing.

File: dvc/external_repo.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (Suor) [reply dvc/external_repo.py]: It should not require changing any tests. This should be a self-contained change.

File: dvc/external_repo.py [hunk 9]
```diff
@@ -77,6 +80,13 @@ def _external_repo(url=None, rev=None, cache_dir=None):
     return new_path
 
 
+def _git_checkout(repo_path, revision):
+    from dvc.scm import Git
+
+    git = Git(repo_path)
+    git.checkout(revision)
```
Comment (Suor) [inline dvc/external_repo.py]: We should probably close is explicitly or we might fail removing it later on Windows.

File: tests/func/test_get.py (hunk 6 above)
Comment (Suor) [reply tests/func/test_get.py]: If you remove `.dvc` and commit that then that won't be a dvc repo :) Now you make a lot of manipulations.

File: tests/func/test_get.py [hunk 10]
```diff
@@ -87,3 +89,43 @@ def test_get_to_dir(dname, erepo):
...
+    storage = git_erepo.mkdtemp()
+
+    dvc_branch = "dvc_test"
+    git_erepo.git.git.checkout("master", b=dvc_branch)
+
+    dvc_repo = Repo.init(git_erepo._root_dir)
```
Comment (Suor) [inline tests/func/test_get.py]: So you make a dvc repo in master branch, looks like this breaks the purpose of the test.

File: dvc/external_repo.py (hunk 9 above)
Comment (mroutis) [reply dvc/external_repo.py]: What do you mean by closing it, @Suor ?

File: tests/func/test_get.py (hunk 10 above)
Comment (pared) [reply tests/func/test_get.py]: As far as I know line 98 is equivalent to:
`git checkout -b dvc_test`

**Comment by efiop** [comment]
@pared Tests are failing, please take a look 🙂

File: dvc/external_repo.py (hunk 9 above)
Comment (efiop) [reply dvc/external_repo.py]: @pared You still need to close it `git.checkout` fails. Add `try&finally`

File: tests/conftest.py [hunk 11]
```diff
@@ -149,21 +151,23 @@ def erepo(repo_dir):
...
-        rconfig.add("upstream", repo.dvc.cache.local.cache_dir, default=True)
+        rconfig.add("upstream", repo.upstream_path, default=True)
         repo.dvc.scm.add([repo.dvc.config.config_file])
         repo.dvc.scm.commit("add remote")
 
         repo.create("version", "master")
         repo.dvc.add("version")
         repo.dvc.scm.add([".gitignore", "version.dvc"])
         repo.dvc.scm.commit("master")
+        repo.dvc.push()
```
Comment (efiop) [inline tests/conftest.py]: Why is this needed? And all upstream_path stuff in general? Sorry, might've missed some discussion somewhere.

Comment (pared) [reply tests/conftest.py]: It seems to me that erepo had default remote configured pointing to its own cache, in order to not trigger infamous NoRemoteError on any test related to `get` and `import`. In new test `test_from_non_dvc_,aster`, where I am removing `.dvc` I had to either manually specify new storage for `erepo`, or adjust `erepo` fixture. I choose the latter because I believe that setting cache as default remote is not realistic use case, which also omits some important details (eg you don't need to push to "add" something to remote cache, because its local cache and adding is done automatically) so we were testing some very specific use cases everywhere where we used `erepo`.

Comment (efiop) [reply tests/conftest.py]: @pared But you are not even using "upstream_path" in your tests. And you have to modify a bunch of unrelated code for no good reason. In your test you simply need `.dvc` to not be tracked on master, which you could do by simply adding `.dvc` to gitignore on master and commiting.

Comment (efiop) [reply tests/conftest.py]: The point about it being not realistic is a good one, but we will soon do that automatically by-default  https://github.com/iterative/dvc/issues/2599 anyway, and, well, it is not nice to break other tests because of 1 test for specific use case.

Comment (pared) [reply tests/conftest.py]: Ok, sorry I didn't noticed that I am modifying repo, its leftover from some previous changes.
Still, I believe this change is desirable. I can achieve the desired `master` state with properly editing `.gitignore` but the actual problem is that erepo is not an actual use case. For example, `test_get_repo_file` should fail in original `erepo` setup. We were using a trick to make it pass, but without the push that we are commenting on, it should not pass.

Comment (pared) [reply tests/conftest.py]: > The point about it being not realistic is a good one, but we will soon do that automatically by-default #2599 anyway, and, well, it is not nice to break other tests because of 1 test for specific use case.

But setting default remote to be a cache is a trick, so in my opinion, it's not breaking other tests, its actually fixing them, since they were not actual use case.

Comment (pared) [reply tests/conftest.py]: #2599 should fix our discussion, but since it's not in the current sprint, I wouldn't just leave it be, especially that this change is not that big.

Comment (efiop) [reply tests/conftest.py]: @pared Sorry, I don't get the point about `test_get_repo_file`, don't see any tricks in it.
```
def test_get_repo_file(erepo):                                                  
    src = erepo.FOO                                                             
    dst = erepo.FOO + "_imported"                                               
                                                                                
    Repo.get(erepo.root_dir, src, dst)                                          
                                                                                
    assert os.path.exists(dst)                                                  
    assert os.path.isfile(dst)                                                  
    assert filecmp.cmp(erepo.FOO, dst, shallow=False)  
```
My point is that you don't need to modify `erepo` and stuff around for your 1 specific test, you can do it with 2 lines that will add `.dvc` dir to gitignore and commit it on master.

File: tests/func/test_get.py [hunk 12]
```diff
@@ -87,3 +89,19 @@ def test_get_to_dir(dname, erepo):
...
+    shutil.rmtree(erepo.dvc.dvc_dir)
+    erepo.dvc.scm.repo.index.remove([erepo.dvc.dvc_dir], r=True)
+    erepo.dvc.scm.commit("remove .dvc")
+
+    caplog.clear()
+    imported_file = "foo_imported"
+    with caplog.at_level(logging.INFO, logger="dvc"):
+        Repo.get(erepo._root_dir, erepo.FOO, out=imported_file, rev="branch")
+
+    assert caplog.text == ""
```
Comment (efiop) [inline tests/func/test_get.py]: Hm, why do we test that there is no info output to logger?

File: tests/conftest.py (hunk 11 above)
Comment (pared) [reply tests/conftest.py]: There is no trick in the test, the trick is inside `erepo`:
```
rconfig = RemoteConfig(repo.dvc.config)
rconfig.add("upstream", repo.dvc.cache.local.cache_dir, default=True)
```

It's not about modifying whole erepo for one test. I believe `erepo` is not being set up properly for tests.

Comment (efiop) [reply tests/conftest.py]: @pared Ah, got it. But that trick is something that we will implement in https://github.com/iterative/dvc/issues/2599 by default, and it doesn't seem to break any other tests, it is just an internal implementation detail for erepo. But I also understand your point about it being hacky. Maybe let's create an issue for it and for now stick with gitignore, so that we don't have to touch all other tests? It doesn't seem like it is in the scope of this PR.

File: tests/func/test_get.py (hunk 12 above)
Comment (pared) [reply tests/func/test_get.py]: https://github.com/iterative/dvc/issues/2858 
I think that we should use `caplog` more often in order to prevent spawning of UI issues that are quite common recently. Do you think I should also check `capsys`?

Comment (efiop) [reply tests/func/test_get.py]: You'll probably get progress bars in capsys, not easy to test, especially automatically, but your call, I don't have a strong opinion here 🙂

File: tests/conftest.py (hunk 11 above)
Comment (pared) [reply tests/conftest.py]: #2888

File: tests/func/test_get.py [hunk 13]
```diff
@@ -87,3 +88,18 @@ def test_get_to_dir(dname, erepo):
 
     assert os.path.isdir(dname)
     assert filecmp.cmp(erepo.FOO, dst, shallow=False)
+
+
+def test_get_from_non_dvc_master(empty_dir, erepo, caplog):
+    erepo.dvc.scm.repo.index.remove([".dvc"], r=True)
```
Comment (efiop) [inline tests/func/test_get.py]: Oh, so you don't even need to gitignore anything 🙂So that erepo change was not needed at all? 🙂

File: tests/func/test_get.py [hunk 14]
```diff
@@ -87,3 +88,18 @@ def test_get_to_dir(dname, erepo):
 
     assert os.path.isdir(dname)
     assert filecmp.cmp(erepo.FOO, dst, shallow=False)
+
+
+def test_get_from_non_dvc_master(empty_dir, erepo, caplog):
```
Comment (efiop) [inline tests/func/test_get.py]: You don't seem to use `empty_dir`, is there really a point in creating a separate fixture for it? Seems like you could simply do your business right in this test.

File: tests/func/test_get.py (hunk 13 above)
Comment (pared) [reply tests/func/test_get.py]: Yep, that's true. :)

File: tests/func/test_get.py (hunk 12 above)
Comment (Suor) [reply tests/func/test_get.py]: pbars go to stderr by default, so shouldn't be an issue.

Comment (Suor) [reply tests/func/test_get.py]: I wonder why `Repo.get()` has not output though.
